Start: Recreate start as a QtWidgets-based app

chennes opened the pull request:
Removes all Start network access and dependency on Chromium for rendering. Written entirely in C++ (no HTML, CSS, JS, etc.).

This started out as a QtQuick tech demo, but it was not possible to cleanly develop QML that worked with both Qt 5.12 and Qt6. That potential migration is left as a future project. The Models developed here were designed for use in QML so should be reusable if that comes to fruition.

Note that this also adds a submodule for the C++ Guidelines Support Library (GSL) that I use to annotate the bare pointers used in QtWidgets development.

----

**WARNING**: This PR now includes code to migrate settings from the old Start workbench to the new Start command. In doing so, references to StartWorkbench and WebWorkbench are removed from your user.cfg file. Consider backing it up if you care about continuing to run Start or Web and preserving your settings.

----

xtemp09 commented:
What is the purpose of engaging QtQuick/QML module? Is FreeCAD going to conquer Android/iOS/Sailfish/HarmonyOS/Fuchsia?


What is the advantage of `gsl::owner` over Qt smart pointers?

yorikvanhavre commented:
I find it excellent to use QtQuick here because QtQuick could help us design more advanced widgets than the ones available by default in Qt. And this is an excellent opportunity to test it in a confined environment first.

Unfortunately, the test seems to fail already :sweat_smile: but that's also a good experience to have.

I'll give this a test!

chennes commented:
> What is the advantage of `gsl::owner` over Qt smart pointers?

GSL isn't a smart pointer, it's an annotation used by static analyzers to track the use of bare pointers. It compiles to nothing. We don't need or want smart pointers here: Qt internally takes ownership of all these things allocated with "new". The use of the GSL here is to tell the linter that we are aware of the ownership, and its transfer.

wwmayer commented on src/Mod/CleanStart/Gui/qml.qrc:
```diff
@@ -0,0 +1,5 @@
+<RCC>
+    <qresource prefix="/">
+        <file>CleanStart.qml</file>
```
@chennes 
Shouldn't this file be part of the PR?

chennes replied:
Thanks for the catch -- I thought I had removed all of the QML, but I forgot about the resources. I'll clean that up.

wwmayer commented on src/Mod/CleanStart/Gui/CleanStartView.cpp:
```diff
@@ -0,0 +1,290 @@
+}
+
+
+void CleanStartView::configureNewFileButtons(QGridLayout* layout)
```
@chennes 
As the linter already reported this causes a build failure because the method is declared as const but in the source file it's not.

wwmayer commented:
Issues:
When loading the workbench then I get these warnings:
```
16:58:59  Cannot read /home/werner/Projects/build_clang/share/examplesQFSFileEngine::open: No file name specified
16:58:59  QIODevice::read (QFile, ""): device not open
16:58:59  QFSFileEngine::open: No file name specified
16:58:59  QIODevice::read (QFile, ""): device not open
16:58:59  QFSFileEngine::open: No file name specified
16:58:59  QIODevice::read (QFile, ""): device not open
16:58:59  QFSFileEngine::open: No file name specified
16:58:59  QIODevice::read (QFile, ""): device not open
16:58:59  QFSFileEngine::open: No file name specified
16:58:59  QIODevice::read (QFile, ""): device not open
```

If in the recent files list there is an image and you load this workbench then the image is displayed in full size instead of the 48x48 px

If a FCStd file doesn't include a thumbnails then the image of a text document is shown. Here the freecad-doc.png icon should be used instead.

If close the new Start Page switch to another workbench and back then FreeCAD crashes.
```
Program received signal SIGSEGV, Segmentation fault.
#0  /lib/x86_64-linux-gnu/libc.so.6(+0x42520) [0x7f25a7d65520]
#1  0x7f25a82a7443 in QtSharedPointer::ExternalRefCountData::getAndRef(QObject const*) from /lib/x86_64-linux-gnu/libQt5Core.so.5+0x53
#2  0x7f25ac6ac183 in QWeakPointer<QObject>::QWeakPointer<QObject, true>(QObject*, bool) from /home/werner/Projects/build_clang/lib/libFreeCADGui.so+0x33
#3  0x7f25ac6ac12a in QWeakPointer<QObject>& QWeakPointer<QObject>::assign<QObject>(QObject*) from /home/werner/Projects/build_clang/lib/libFreeCADGui.so+0x2a
#4  0x7f25acd88171 in QPointer<Gui::MDIView>::operator=(Gui::MDIView*) from /home/werner/Projects/build_clang/lib/libFreeCADGui.so+0x21
#5  0x7f25acd79be4 in Gui::MainWindow::setActiveWindow(Gui::MDIView*) from /home/werner/Projects/build_clang/lib/libFreeCADGui.so+0x74
#6  0x7f256ca35d19 in CleanStartGui::Workbench::loadCleanStart() from /home/werner/Projects/build_clang/Mod/CleanStart/CleanStartGui.so+0xb9
#7  0x7f256ca35c51 in CleanStartGui::Workbench::activated() from /home/werner/Projects/build_clang/Mod/CleanStart/CleanStartGui.so+0x11
#8  0x7f25ac4a099c in Gui::Application::activateWorkbench(char const*) from /home/werner/Projects/build_clang/lib/libFreeCADGui.so+0x168c
#9  0x7f25ac531844 in Gui::Application::sActivateWorkbenchHandler(_object*, _object*) from /home/werner/Projects/build_clang/lib/libFreeCADGui.
```

chennes commented:
Another issue I just noticed -- if the Examples directory contains things like text documents (or cMake files...), those are displayed. So I need to filter that list.

chennes commented:
@wwmayer I believe I have addressed the issues you pointed out (though my Windows build is giving me some difficulty today, so if I missed something please let me know). I noticed that freecad-doc.svg exists, but was not being compiled into our resources, so I have added it there (so I can scale the icon to larger sizes).

EmperorArthur commented:
> What is the purpose of engaging QtQuick/QML module? Is FreeCAD going to conquer Android/iOS/Sailfish/HarmonyOS/Fuchsia?

Popping in as a random user.  While I love the power of C++, writing a GUI programmatically can turn into a pain maintenance wise. One of the major advantages of QML is separation of concerns.  It allows modifying the front-end without recompiling, and opens development up to people who don't know C++.  

The other question is, are you sure everything in FreeCAD works on all common screen resolutions / window sizes?  Because while QML doesn't solve that problem it allows a graphics person to work on the issue without having to learn C++.

adrianinsaval commented:
Do we need to keep this as a workbench? Can't we just insert a tab with the UI without registering a workbench for this? At most we should have command to bring up the startpage again

yorikvanhavre commented:
i'd agree with that! module, yes, workbench, no :) it could even bring interesting ideas, such as loading a workbench while you are  looking at the start screen...

chennes commented:
Yes, that's a good idea -- I'll make that change.

RohanMishra315 commented:
https://github.com/FreeCAD/FreeCAD/issues/12565#issuecomment-2027983292
Hey @chennes I have been contributing to FreeCad for past 6 months , you can look at all my commits. As this issue had been selected as Gsoc project I had been working on it for past 2 weeks like creating mockups desings for QT workbench and also working on the proposal . But, now these has been removed from Gsoc project , I feel kinda bad and gutted but at the same time glad that a work as started on this issue. Even then , I really want to work in it , if there is some work are still there that it can still be used as Gsoc project , I'll be glad to work on it. 
Thank you

RohanMishra315 commented:
@chennes

chennes commented:
There will always still be work to do. This started to take higher priority for us so that we could remove the QtWebEngine dependency (which was problematic for a number of reasons). As you can see, the core of Start is a very tiny chunk of code that is just a few buttons. From your mockup on #12565 it seems you have larger ideas for Start -- if you feel that's a large enough project for GSoC then by all means continue work on it.

RohanMishra315 commented:
@chennes yeah from the #12565 discussion we can a lot of things in Start page and it can be a Gsoc level project . So, is this still open for Gsoc , should I make proposal for it.

chennes commented:
I'm concerned about your statement that you are "not good at design" -- the whole Start project is design, the code is trivial. So I am not sure it's appropriate. Have you reached out to the Design Working Group on the FreeCAD Discord?

PaddleStroke commented:
+1 to not have a workbench. Let's just keep one Tab in the document tab. And this tab should not have the X button so that it cannot be closed. Also it does not need text, it could only be a 'home' icon.

chennes commented:
I think not making it a workbench is fine, but I don't know how to do that. How do I create a Python command (called from InitGui.py) that will run the command-creation code? I've only ever done that with a Workbench.

PaddleStroke commented:
Perhaps you can look at the Ondsel addon, it's adding a dock widget but not a workbench : https://github.com/Ondsel-Development/Ondsel-Lens
There are few other addons that are not adding a workbench, such as triplus tab bar.

yorikvanhavre commented:
> And this tab should not have the X button so that it cannot be closed. Also it does not need text, it could only be a 'home' icon.

Why should it be unclosable? i think one might want to get rid of it... I think having a command that reopens it (in Tools menu for ex) is sufficient

PaddleStroke commented:
Exactly to avoid having a command to reopens it.
Adding commands for every thing is not a good thing, it adds complexity and leads to unecessary hassle. Say a casual user close the home page by mistake and want it back. Well he has to look for how to get it back online, then look for the command etc.

And what use case is there to close the home page ?
- User don't want to use the start page at all : for that a preference to disable home page is better.
- User has too many documents and he wants more space : This is why I said only a icon and no text like so that it takes no space : 
![image](https://github.com/FreeCAD/FreeCAD/assets/19984177/d08955a0-cfbf-4974-98a9-f2dd98f7b163)

yorikvanhavre commented:
i still think one might want sometimes one and sometimes the other, but i have no strong opinion.

wwmayer commented:
> And this tab should not have the X button so that it cannot be closed.

I don't see a reason either why it shouldn't be closable. In fact whenever I start with a clean config the first thing I do is to close the Start page.

IMO, if there is a command in the Help menu to re-open it that would be sufficient.

> I think not making it a workbench is fine, but I don't know how to do that. How do I create a Python command (called from InitGui.py) that will run the command-creation code? I've only ever done that with a Workbench.

Some time ago I added the possibility to let a module inject commands to menus or toolbars of workbenches. This was needed to reduce cross-dependencies between the modules.

All what you need is to add a workbench manipulator class that inherits from `Gui::WorkbenchManipulator`. There you can override the methods modifyMenuBar() or modifyToolBars() to insert the command into the MenuItem or ToolBarItem tree.

Then you have to register the manipulator class inside `PyMOD_INIT_FUNC(StartGui)`

In the Part and Material modules you will find some example code how to achieve this.

chennes commented:
> All what you need is to add a workbench manipulator class that inherits from `Gui::WorkbenchManipulator`.

Perfect, that's exactly what I needed. I've migrated Start to being a simple single command, "Start_Start". Right now I'm just appending it to the Help menu. Any opinions on where it should be in the order?

There will be at least one more commit in this PR, because now I need to migrate the hardcoded Start stuff in core to use the command instead of the workbench.

chennes commented:
In addition to deciding where to place that preference, how should we go about actually running the Start command on startup? This used to be handled by setting it as the AutoloadWorkbench, but now that it is just a command, some code needs to execute it, pending the value of that user parameter. Is there already a mechanism in place to enable this sort of behavior? i.e. "Run all of these commands on startup" or something along those lines?

chennes commented:
Another suggestion from M3 - checkbox on Start page itself to fire off start or not.

chennes commented:
This is close to working correctly: as far as I can determine there is only one blocker to the merge. FreeCAD does not behave correctly with the startup workbench set to "NoneWorkbench" (which I was hoping to make the new default, now that Start is not a WB). No toolbars show and the menus are almost unpopulated. This also means that clicking the "New Empty File" button yields and awkward interface that requires you to choose a workbench from the View menu before you can do anything. I'm not sure what the right path forward here is. @wwmayer is there a better default choice than "NoneWorkbench"? Or a way to trigger creation of the global menus and toolbars even if NoneWorkbench is activated? 

(Note that the failing unit test is because NoneWorkbench is already active, so trying to activate it returns False, which is an unexpected result -- the unit test relies on the first workbench in the WBs list NOT being active at the time it is run)

PaddleStroke commented:
> In the first place FreeCAD is a CAD program. So why not using PartDesign or Part as the default startup workbench?

+1
I think if you make a poll, more than 80% will want default as PartDesign.

wwmayer commented:
@chennes 
With 12857cc249 the entire Web module will be removed. But the Web module contains a simple server that allows a TCP/IP connection to control FreeCAD from the remote. That's a functionality some people have asked for in the past:
* https://forum.freecad.org/viewtopic.php?t=13420
* https://forum.freecad.org/viewtopic.php?t=59724
* https://forum.freecad.org/viewtopic.php?t=73257
* https://forum.freecad.org/viewtopic.php?t=59669
* https://forum.freecad.org/viewtopic.php?t=50647

chennes commented:
Thanks for the feedback: I will make PartDesign the default startup WB. I will also drop 12857cc24948cecac77574167df38a13e7b27aa1 and replace it with something that leaves the server in place, and only removes the web browsing component.

chennes commented:
In my testing of this change, I am running into a problem where the File and Workbench toolbars are being disabled: I re-enable them, save, close, and restart, and they switch back to disabled. I'm not seeing the flaw in the code, so I've probably been staring at it too long. Anyone else see why that's happening? (or alternately, is it not happening on your system?)

FEA-eng commented:
So with this PR merged what the user will see by default when starting FreeCAD is this, right?

![open](https://github.com/FreeCAD/FreeCAD/assets/59876896/05ca6de9-a64c-4253-88d8-f051da874b74)

The problem is that I have to click the small New button so that everything (including the 3D viewport) is not grayed out anymore. If this is what it will look like now (and not just because my config file is not fresh or the works is not completed yet) then I think that there should be some window with big New and Open buttons (or just the Start page shown by default?) because it's really confusing that you have to find that small New button to activate everything and there's no indication that it should be done. New users will think that FreeCAD wasn't started properly.

Also, will the Start page get its style back after the announced further changes or will it look like this from now on?

![start](https://github.com/FreeCAD/FreeCAD/assets/59876896/25b6f426-3b11-451c-b4ab-e50b7f88bb0f)

The previous style (after improvements in 0.22) was much better - more modern.

PaddleStroke commented:
I'm having some build errors after rebasing on main: 

> Severity	Code	Description	Project	File	Line	Suppression State
> Error	C1083	Cannot open include file: '3rdParty/GSL/include/gsl/pointers': No such file or directory	StartGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\Gui\StartView.cpp	47	
> Error	MSB8066	Custom build for 'C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\TestStart\__init__.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\TestStart\TestStartPage.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\Init.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\TestStartGui.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\InitGui.py' exited with code 1.	StartTests	C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets	242	
> Error	MSB8066	Custom build for 'C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\StartPage.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\TranslationTexts.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\__init__.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\LoadMRU.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\LoadExample.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\LoadNew.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\LoadCustom.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\StartPage.css;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\StartPage.js;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\StartPage.html;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\EnableDownload.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\OpenSettings.py;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\userhub.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\poweruserhub.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\developerhub.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\manual.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\freecad.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\installed.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_file_thumbnail.svg;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\icon_settings.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_assembly.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_empty_file.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_open_file.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_parametric_part.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_csg_part.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_2d_draft.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\new_architecture.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\icon_documents.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\icon_help.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\icon_activity.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\images\icon_blog.png;C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\StartPage\CMakeLists.txt' exited with code 1.	StartPage	C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets	242	
> Error	C1083	Cannot open include file: '3rdParty/GSL/include/gsl/pointers': No such file or directory	Start	C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\App\AppStart.cpp	32	
> Error	MSB3073	The command "setlocal
> cd "C:\Users\Sky_l\Desktop\FC build\src\Mod\Web\Gui"
> if %errorlevel% neq 0 goto :cmEnd
> C:
> if %errorlevel% neq 0 goto :cmEnd
> "C:\Program Files\CMake\bin\cmake.exe" -E cmake_autogen "C:/Users/Sky_l/Desktop/FC build/src/Mod/Web/Gui/CMakeFiles/WebGui_autogen.dir/AutogenInfo.json" Release
> if %errorlevel% neq 0 goto :cmEnd
> :cmEnd
> endlocal & call :cmErrorLevel %errorlevel% & goto :cmDone
> :cmErrorLevel
> exit /b %1
> :cmDone
> if %errorlevel% neq 0 goto :VCEnd
> :VCEnd" exited with code 1.	WebGui	C:\Program Files\Microsoft Visual Studio\2022\Community\MSBuild\Microsoft\VC\v170\Microsoft.CppCommon.targets	146	
> Error	C1083	Cannot open include file: '3rdParty/GSL/include/gsl/pointers': No such file or directory	StartGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\Gui\FileCardDelegate.cpp	41	
> Error	C1083	Cannot open include file: '3rdParty/GSL/include/gsl/pointers': No such file or directory	StartGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\Gui\Manipulator.cpp	39	
> Error	C1083	Cannot open include file: '3rdParty/GSL/include/gsl/pointers': No such file or directory	StartGui	C:\Users\Sky_l\Desktop\Freecad-Src\src\Mod\Start\Gui\AppStartGui.cpp	38

PaddleStroke commented:
Is there something that needs to be done about this GSL 3rd party?

PaddleStroke commented:
I'll try git submodule update --init and rebuild.

PaddleStroke commented:
I do not understand, I don't see the start page anymore, it's just gone, the workbench is gone too.
@FEA-eng you mention a 'small button' but I cannot find this button. Also in your screenshot I do not see it neither.
Where is it?

If I cannot find it after 10 minutes of searching, I think it is safe to say that most won't find it neither. So the start page is as good as gone in current state.

FEA-eng commented:
@FEA-eng you mention a 'small button' but I cannot find this button. Also in your screenshot I do not see it neither. Where is it please?

I’m talking about the New button on the toolbar (top left, right under the File menu in my first screenshot).

obelisk79 commented:
I don't see the new start workbench anywhere, I just fresh compiled to confirm I hadn't missed an important commit but now I get nothing.

Rexbas commented:
Same here. Start is also not shown anymore in the workbenches drop down menu and now defaults to Part Design?! I don't know if this intended. When I switch workbenches I do get this error every time, no matter what I select.
`<Command> Command.cpp(1841): duplicate command Start_Start`

furgo16 commented:
>  Start is also not shown anymore in the workbenches drop down menu and now defaults to Part Design

If I understood it correctly, this was a design decision: https://github.com/FreeCAD/FreeCAD/pull/13134#issuecomment-2066579022

Roy-043 commented:
With the removal of the Web WB Std_WhatsThis no longer works OOTB. The easiest fix is to only allow the `show_browser` option:

https://github.com/FreeCAD/FreeCAD/blob/eb3e83b97a345620e0e26f3051d3d7a3f26755c6/src/Mod/Help/Help.py#L112-L114

```
OS: Windows 8 build 9600
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.36958 (Git)
Build type: Release
Branch: main
Hash: a041129090b8dcbb8367636840d3c668ef30a933
Python 3.11.9, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: Dutch/Netherlands (nl_NL)
```
